localstack/services/stepfunctions/asl/component/state/state_execution/state_task/service/state_task_service_callback.py

```diff
@@ -46,8 +47,35 @@ def _eval_service_task(self, env: Environment, parameters: dict):
     def _wait_for_task_token(self, env: Environment) -> None:  # noqa
         callback_id = env.context_object_manager.context_object["Task"]["Token"]
         callback_endpoint = env.callback_pool_manager.get(callback_id)
-        outcome = callback_endpoint.wait()  # TODO: implement timeout.
 
+        # With Timeouts-only definition:
+        if not self.heartbeat:
+            self.timeout.eval(env=env)
+            timeout_seconds = env.stack.pop()
+            # Although the timeout is handled already be the superclass (ExecutionState),
+            # the timeout value is specified here too, to allo this child process to terminate earlier even if
+            # discarded by the main process.
+            # Note: although this is the same timeout value, this can only decay strictly after the first timeout
+            #       started as it is invoked strictly later.
+            outcome = callback_endpoint.wait(timeout=timeout_seconds)
+        else:
+            self.heartbeat.eval(env=env)
+            heartbeat_seconds = env.stack.pop()
+            heartbeat_endpoint: HeartbeatEndpoint = callback_endpoint.setup_heartbeat_endpoint(
+                heartbeat_seconds=heartbeat_seconds
+            )
+
+            outcome = None
+            while (
+                env.is_running() and outcome is None
+            ):  # Until subprocess hasn't timed out or result wasn't received.
+                received = heartbeat_endpoint.clear_and_wait()
+                if not received and env.is_running():  # Heartbeat timed out.
+                    raise TimeoutError()
+                outcome = callback_endpoint.get_outcome()
+
+        if outcome is None:
+            raise TimeoutError()
```

Comment:
Indeed, thank you. We have 4 types of TimeoutError: due to waitForTaskToken, heartbeat and task timeout. These are all handled by the same logic (timeout error) hence why I never felt the need to differentiate. This may change in the future though. I have added 3 subtypes of TimeoutError to the implementation, this should improve on the situation.